src/NodeValueReader.h

```diff
@@ -0,0 +1,168 @@
+#ifndef NODE_VALUE_READER_H
+#define NODE_VALUE_READER_H
+
+#if !defined (ACE_LACKS_PRAGMA_ONCE)
+#pragma once
+#endif /* ACE_LACKS_PRAGMA_ONCE */
+
+#include <dds/DCPS/ValueReader.h>
+
+#include <nan.h>
+
+namespace NodeOpenDDS {
+
+template <typename T, typename U>
+void safe_assign(T& out, const U& in)
+{
+  out = static_cast<T>(in);
+}
+
+template <typename U>
+void safe_assign(ACE_CDR::LongDouble& out, const U& in)
+{
+  ACE_CDR_LONG_DOUBLE_ASSIGNMENT(out, in);
+}
+
+class NodeValueReader : public OpenDDS::DCPS::ValueReader {
+public:
+  explicit NodeValueReader(v8::Local<v8::Object> obj);
+
+  bool begin_struct();
+  bool end_struct();
+  bool begin_struct_member(OpenDDS::XTypes::MemberId& member_id, const OpenDDS::DCPS::MemberHelper& helper);
+  bool end_struct_member();
+
+  bool begin_union();
+  bool end_union();
+  bool begin_discriminator();
+  bool end_discriminator();
+  bool begin_union_member();
+  bool end_union_member();
+
+  bool begin_array();
+  bool end_array();
+  bool begin_sequence();
+  bool elements_remaining();
+  bool end_sequence();
+  bool begin_element();
+  bool end_element();
+
+  bool read_boolean(ACE_CDR::Boolean& value);
+  bool read_byte(ACE_CDR::Octet& value);
+#if OPENDDS_HAS_EXPLICIT_INTS
+  bool read_int8(ACE_CDR::Int8& value);
+  bool read_uint8(ACE_CDR::UInt8& value);
+#endif
+  bool read_int16(ACE_CDR::Short& value);
+  bool read_uint16(ACE_CDR::UShort& value);
+  bool read_int32(ACE_CDR::Long& value);
+  bool read_uint32(ACE_CDR::ULong& value);
+  bool read_int64(ACE_CDR::LongLong& value);
+  bool read_uint64(ACE_CDR::ULongLong& value);
+  bool read_float32(ACE_CDR::Float& value);
+  bool read_float64(ACE_CDR::Double& value);
+  bool read_float128(ACE_CDR::LongDouble& value);
+
+#ifdef NONNATIVE_LONGDOUBLE
+  bool read_float128(long double& value);
+#endif
+
+  bool read_fixed(OpenDDS::FaceTypes::Fixed& value);
+  bool read_char8(ACE_CDR::Char& value);
+  bool read_char16(ACE_CDR::WChar& value);
+  bool read_string(std::string& value);
+  bool read_wstring(std::wstring& value);
+
+  bool read_long_enum(ACE_CDR::Long& value, const OpenDDS::DCPS::EnumHelper& helper);
+
+private:
+  NodeValueReader();
+
+  std::string indent(int mod);
+
+  bool begin_nested();
+  bool end_nested();
+
+  template <typename V, typename T>
+  bool primitive_helper(T& value, bool (v8::Value::*checker)() const)
```

Comment:
Using typedefs to give these "pointer-to-member function" and function pointer types their own names may improve readability (here and below).

Reply:
It looks like there's more of these that could change.  Doesn't need to happen right away this PR.

Reply:
I'm not sure the str_conv ones can get a simple typedef because their return types depend on a template parameter.